### PR TITLE
doc: Add Shell-Related documentation

### PR DESCRIPTION
**Pull request by ubieda**
### Description
Add Shell-Related Documentation:
- General
    - Add Backend configuration options for USB and Bluetooth LE.
    - Add Configuration for commonly-used Shell command groups (GPIO, I2C, Sensors, Flash, File-System).
- Shell module sample
    - Add first version of README describing different types of options registering Shell commands.

### CODE REVIEW

File: samples/subsys/shell/shell_module/README.rst [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+exposing the shell interface over Bluetooth LE (NUS). The following commands
+are applicable to the nrf52840dk/nrf52840.
+
+.. zephyr-app-commands::
+   :zephyr-app: samples/subsys/shell/shell_module
+   :host-os: unix
+   :board: nrf52840dk/nrf52840
+   :goals: flash
+   :gen-args: -DOVERLAY_CONFIG='overlay-bt.conf' -DDTC_OVERLAY_FILE='bt.overlay'
+   :compact:
```
Comment (kartben) [inline samples/subsys/shell/shell_module/README.rst:117]: mmm shouldn't we just encourage people to use existing snippets and drop these overlays? They seem to be mostly a copy-paste of the equivalent snippets. I think these docs on "how to enable CDC/NUS/RTT shell" should be moved to the actual shell documentation page, as they have nothing to do with this sample specifically (even more so if the generic snippets can be used)

File: samples/subsys/shell/shell_module/README.rst [hunk 2]
```diff
@@ -0,0 +1,157 @@
+.. zephyr:code-sample:: shell-module
+   :name: Shell Sample
+   :relevant-api: shell_api
+
+   Basic sample for utilizing the Zephyr Shell
```
Comment (kartben) [inline samples/subsys/shell/shell_module/README.rst:5]:
```suggestion
.. zephyr:code-sample:: shell-module
   :name: Custom shell module
   :relevant-api: shell_api

   Register shell commands using the Shell API
```

**Review by kartben — CHANGES_REQUESTED** [review]
Thanks for adding this README - this will be one more to tick off of https://github.com/zephyrproject-rtos/zephyr/issues/27805 :)

However, this will likely need more work as the goal of the shell module sample is actually to demonstrate how to use the shell API for registering new commands, sub-commands, etc.

I like the idea of documenting somewhere the list of useful shell modules (GPIO, I2C, ...) and their associated KConfigs, but I don't think this belongs to this README.

Would you be able to rework this page so that it's indeed documenting the actual "custom shell module" aspects of it, and maybe move the instructions on how to enable CDC-ACM, NUS, etc. to the shell doc page?

Thanks!

**Review by alwa-nordic — DISMISSED** [review]

**Review by jori-nordic — DISMISSED** [review]
Thanks! I'm always looking for how to enable the various extra shell commands. 

I agree with @kartben that the various backends maybe should be in a more general shell page and linked in the README.
But not nacking, we can always move that later.

**Comment by kartben** [comment]
> Thanks! I'm always looking for how to enable the various extra shell commands.

But again, this is *not* what this sample is meant for :) If we need (and we do, I think we all agree!) something like "Shell tips and tricks" then this belongs to the shell documentation, not the sample that's meant to explain how to actually write one's own shell commands.

**Comment by ubieda** [comment]
@kartben @jori-nordic Thanks for your feedback - I agree with @kartben, I'll refactor this PR to capture those "tips and tricks" under the Shell documentation, and have the sample README expand on what the sample is doing.

**Comment by ubieda** [comment]
@kartben sorry for taking a bit to iterate on this. Expect an update before the week ends!

**Comment by ubieda** [comment]
I've made a second pass of this PR, structuring the shell-related documentation in two sections:
- One on the existing Shell documentation, where details are captured on how to configure backends and commonly-used Shell commands.
- One on the Shell module sample, where some content regarding the registration of Shell commands is summarized.

However, as I was going through the documentation on the Shell sample, I realize the existing Shell documentation kind of captures most of the details on registering the different types of Shell commands. I did a first effort to capture a list on the README while avoiding content duplication, but there may be room for improvement, as well as desire to go a bit more in-depth in certain aspects.

I'm open to feedback as to what should go where and what could use more detail. As mentioned, this is another pass and could have more iterations.

File: doc/services/shell/index.rst [hunk 3]
```diff
@@ -97,6 +99,86 @@ considerably. For that cost, it will enable the line editing,
 `tab completion <tab-feature_>`_, and `history <history-feature_>`_
 features of the shell.
 
+USB CDC ACM Backend
+===================
+
+To configure Shell USB CDC ACM backend:
+
+* Enable :kconfig:option:`CONFIG_SHELL_BACKEND_SERIAL`, along with other USB
```
Comment (jori-nordic) [inline doc/services/shell/index.rst:107]: I thought we had a snippet?

File: samples/subsys/shell/shell_module/README.rst [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+      log-level  : <module name> <severity (0-4)>
+    uart:~$ kernel version
+    Zephyr version 3.6.99
+    uart:~$
+
+Details on Shell Subsystem
+==========================
+
+For more details on the Shell subsystem, check the general Shell documentation:
+:zephyr_file:`doc/services/shell/index.rst`
```
Comment (jori-nordic) [inline samples/subsys/shell/shell_module/README.rst:200]: does linking to the rst section work from samples? I don't remember.
It's either sample -> general doc or general doc -> sample linking that doesn't.

File: doc/services/shell/index.rst (hunk 3 above)
Comment (ubieda) [reply doc/services/shell/index.rst:107]: If I understand it correctly, the snippet cdc-acm-console is missing the zephyr,shell-uart redirection to the appropriate node.

I can file a separate PR to have that extra line included in the snippet.

LMk if I'm missing something!

Comment (jori-nordic) [reply doc/services/shell/index.rst:107]: ah alright, I was thinking of the console then. disregard the comment

File: doc/services/shell/index.rst [hunk 5]
```diff
@@ -97,6 +99,86 @@ considerably. For that cost, it will enable the line editing,
 `tab completion <tab-feature_>`_, and `history <history-feature_>`_
 features of the shell.
 
+USB CDC ACM Backend
+===================
```
Comment (kartben) [inline doc/services/shell/index.rst:103]: it looks like backend "headings" are a bit all over the place (see USB CDC, Telnet, RTT, ...) can you have a look and see if at a minimum you can't group USB CDC and Telnet together under the same heading (and maybe move RTT there too, if appropriate?)

**Review by kartben — CHANGES_REQUESTED** [review]

File: doc/services/shell/index.rst (hunk 5 above)
Comment (ubieda) [reply doc/services/shell/index.rst]: I've updated this branch attempting to gather all the already described backends. However, the list is not exhaustive and did not try to group using Shell as Logging backend: hopefully that's okay!

File: samples/subsys/shell/shell_module/README.rst (hunk 1 above)
Comment (ubieda) [reply samples/subsys/shell/shell_module/README.rst]: I've refactored the Shell documentation to reference the code-snippets (now shell-compatible for both `cdc-acm-console`, as well as for `nus-console`)

File: samples/subsys/shell/shell_module/README.rst (hunk 4 above)
Comment (ubieda) [reply samples/subsys/shell/shell_module/README.rst]: I've changed it to **:ref:`Shell documentation <shell_api>`** and it does appear to work on a local build.

File: doc/services/shell/index.rst [hunk 6]
```diff
@@ -112,6 +194,46 @@ types:
...
+
+	CONFIG_FLASH=y
+	CONFIG_FLASH_SHELL=y
+
+* File-System
+
+.. code-block:: kconfig
+
+	CONFIG_FILE_SYSTEM=y
+	CONFIG_FILE_SYSTEM_SHELL=y
```
[the remark this answers is not in the document]
Comment (ubieda) [reply doc/services/shell/index.rst]: I've gotten rid of these code-blocks altogether. Instead, I've changed it to a list and made the options linkable to their Kconfig source.

File: samples/subsys/shell/shell_module/README.rst [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+
+      /* Enable obscure mode */
+      shell_obscure_set(sh, true);
+
+      /* User-input is replaced with asterisks */
+      ...
+
+      /* Disable obscure mode */
+      shell_oscure_set(sh, false);
+      ...
```
[the remark this answers is not in the document]
Comment (ubieda) [reply samples/subsys/shell/shell_module/README.rst]: Done!

File: doc/services/shell/index.rst [hunk 8]
```diff
@@ -97,6 +99,86 @@ considerably. For that cost, it will enable the line editing,
...
+	# Serial Shell Backend
+	CONFIG_SHELL_BACKEND_SERIAL=y
+	# USB Device configuration
+	CONFIG_USB_DEVICE_STACK=y
+	CONFIG_USB_DEVICE_PRODUCT="Zephyr USB shell sample"
+	CONFIG_SHELL_BACKEND_SERIAL_CHECK_DTR=y
+	CONFIG_UART_LINE_CTRL=y
+
+* Assign the chosen property `zephyr,shell-uart` to a node compatible with
+  `zephyr,cdc-acm-uart`:
```
[the remark this answers is not in the document]
Comment (ubieda) [reply doc/services/shell/index.rst]: Done!

File: samples/subsys/shell/shell_module/README.rst [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+      cycles     : Kernel cycles.
+      stacks     : List threads stack usage.
+      threads    : List kernel threads.
+      uptime     : Kernel uptime. Can be called with the -p or --pretty options
+      version    : Kernel version.
+      sleep      : ms
+      log-level  : <module name> <severity (0-4)>
+    uart:~$ kernel version
+    Zephyr version 3.6.99
+    uart:~$
```
[the remark this answers is not in the document]
Comment (ubieda) [reply samples/subsys/shell/shell_module/README.rst]: I agree - Done!

File: samples/subsys/shell/shell_module/README.rst [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+   :name: Custom Shell module
+   :relevant-api: shell_api
+
+   Register shell commands using the Shell API
+
+Overview
+********
+
+This is a simple application demonstrating how to write and register commands
+using the Shell API.
```
[the remark this answers is not in the document]
Comment (ubieda) [reply samples/subsys/shell/shell_module/README.rst]: Done!

File: samples/subsys/shell/shell_module/README.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (ubieda) [reply samples/subsys/shell/shell_module/README.rst]: Done!

File: doc/services/shell/index.rst [hunk 11]
```diff
@@ -97,6 +99,86 @@ considerably. For that cost, it will enable the line editing,
...
+==========================
+
+To configure Bluetooth LE (NUS) backend:
+
+* Enable :kconfig:option:`CONFIG_SHELL_BACKEND_SERIAL`, along with the other
+  Bluetooth LE configurations:
+
+.. code-block:: kconfig
+
+	# Enable Serial
```
[the remark this answers is not in the document]
Comment (ubieda) [reply doc/services/shell/index.rst]: Done - Hopefully I didn't miss anything

File: doc/services/shell/index.rst (hunk 3 above)
Comment (ubieda) [reply doc/services/shell/index.rst]: Integrated now!

**Review by alwa-nordic — DISMISSED** [review]

**Review by jakub-uC — DISMISSED** [review]
Thank you

File: samples/subsys/shell/shell_module/README.rst [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+   Register shell commands using the Shell API
+
+Overview
+********
+
+This is a simple application demonstrating how to write and register commands
+using the :ref:`Shell API <shell_api>`:
+
+Register Static commands
+   ```version``` is a static command that prints the kernel version.
```
Comment (kartben) [inline samples/subsys/shell/shell_module/README.rst:14]: double backquotes, not triple, please

image

File: doc/services/shell/index.rst [hunk 13]
```diff
@@ -49,10 +54,76 @@ At this point, the following transport layers are implemented:
...
+similar to a UART backend in that regard. This will make the client send a character
+as soon as it is typed having the effect of increasing the network traffic
+considerably. For that cost, it will enable the line editing,
+`tab completion <tab-feature_>`_, and `history <history-feature_>`_
+features of the shell.
+
+USB CDC ACM
+===========
+
+To configure Shell USB CDC ACM backend, simply add the snippet `cdc-acm-console`
```
Comment (kartben) [inline doc/services/shell/index.rst:86]:
```suggestion
To configure Shell USB CDC ACM backend, simply add the snippet ``cdc-acm-console``
```

File: doc/services/shell/index.rst [hunk 14]
```diff
@@ -49,10 +54,76 @@ At this point, the following transport layers are implemented:
...
+
+Details on the configuration settings are captured in the following files:
+
+- :zephyr_file:`snippets/cdc-acm-console/cdc-acm-console.conf`.
+- :zephyr_file:`snippets/cdc-acm-console/cdc-acm-console.overlay`.
+
+Bluetooth LE (NUS)
+==================
+
+To configure Bluetooth LE (NUS) backend, simply add the snippet `nus-console`
```
Comment (kartben) [inline doc/services/shell/index.rst:101]:
```suggestion
To configure Bluetooth LE (NUS) backend, simply add the snippet ``nus-console``
```

**Review by kartben — CHANGES_REQUESTED** [review]
this is brilliant, thanks @ubieda!
Some backticks fun and it should be good to go

File: samples/subsys/shell/shell_module/README.rst (hunk 12 above)
Comment (ubieda) [reply samples/subsys/shell/shell_module/README.rst]: Roger - got confused by this suggestion: https://github.com/zephyrproject-rtos/zephyr/pull/71904#discussion_r1603638630.

Addressed.

Comment (kartben) [reply samples/subsys/shell/shell_module/README.rst]: > Roger - got confused by this suggestion: [#71904 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/71904#discussion_r1603638630).

woops, sorry!

**Review by kartben — APPROVED** [review]

**Review by simhein — APPROVED** [review]
Thanks, LGTM

**Pull request merged**